Refactor Block Publisher to support Consensus Engine prototyping

aludvik opened the pull request:
This PR makes some initial strides towards supporting the consensus engine api, although it does not get all the way there. It also cleans up the block publisher code to make it easier to understand when this is easy to do.

agunde406 commented on validator/sawtooth_validator/journal/publisher.py:
```diff
@@ -556,24 +556,24 @@ def get_current_queue_info(self):
         """ Build a candidate block and construct the consensus object to
         validate it.
-        :param chain_head: The block to build on top of.
+        :param previous_block_id: The block to build on top of.
```
previous_block_id -> previous_block

agunde406 commented on validator/sawtooth_validator/journal/publisher.py:
```diff
@@ -486,9 +486,7 @@ def __init__(self,
         self._block_sender = block_sender
         self._batch_publisher = BatchPublisher(identity_signer, batch_sender)
-        self._pending_batches = []  # batches we are waiting for validation,
-        # arranged in the order of batches received.
```
Typo in corresponding commit message: "using a three separate" ->  "using three separate"

agunde406 commented on validator/sawtooth_validator/journal/publisher.py:
```diff
@@ -812,14 +765,81 @@ def on_check_publish_block(self, force=False):
+    def rebuild(self, committed=None, uncommitted=None):
+        """Recomputes the list of pending batches
+
+        :param committed_batches: Batches committed in the current chain
```
Param names have changed, update docstring.

nick-drozd commented on validator/sawtooth_validator/journal/publisher.py:
```diff
@@ -812,14 +765,81 @@ def on_check_publish_block(self, force=False):
+        # Ordered batches waiting to be processed
+        self._batches = list()
+        self._ids = set()
+        self._gauge = COLLECTOR.gauge('pending_batch_gauge', instance=self)
```
In the current pattern, all the metrics are defined in the same place. Would it be better to leave the gauge in the publisher and pass it in as an argument?

nick-drozd commented on validator/sawtooth_validator/journal/publisher.py:
```diff
@@ -812,14 +765,81 @@ def on_check_publish_block(self, force=False):
+
+        self.reset()
+
+        # Uncommitted and pending disjoint sets
```
are disjoint sets

peterschwarz commented on validator/sawtooth_validator/journal/publisher.py:
```diff
@@ -812,14 +765,81 @@ def on_check_publish_block(self, force=False):
         return False
 
 
+class PendingBatchesPool:
```
I suppose "Pool" is a bit more accurate than `Queue`

nick-drozd commented:
Commit message for `Refactor pending batches tracking into class`: previously done using __a__ three
separate data structures

peterschwarz commented on validator/sawtooth_validator/journal/publisher.py:
```diff
@@ -833,6 +833,7 @@ def update(self, still_pending, last_sent):
         last_index = self._batches.index(last_sent)
         unsent = self._batches[last_index + 1:]
         self._batches = still_pending + unsent
+        self._ids = set(b.header_signature for b in self._batches)
```
Why not just replace `self._ids` with a generator?  Seems like it would be better to use `self._batches` as the source of truth, and the ids are just a view.  Secondly, making it a set where as self._batches is a list seems unnecessary or is missing a check that the should be same length - any discrepancy implies there is a duplicate batch in there.

aludvik replied:
The main benefit of using a set like this is improved lookup times. You are correct to point out that it does not correctly handle duplicate batches though, so I will add a check around that.

aludvik commented on validator/sawtooth_validator/journal/publisher.py:
```diff
@@ -52,6 +53,24 @@
+
+
+class NoPendingBatchesRemaining(Exception):
+    """There are pending batches remaining to be processed."""
```
typo s/are/are not/

agunde406 commented on validator/sawtooth_validator/journal/publisher.py:
```diff
@@ -59,6 +59,18 @@ class ConsensusNotReady(Exception):
     """
 
 
+class NoPendingBatchesRemaining(Exception):
```
The name seems different from the message

peterschwarz commented on validator/sawtooth_validator/journal/publisher.py:
```diff
@@ -744,12 +742,30 @@ def has_batch(self, batch_id):
+        self._ids = ConcurrentSet()
+
+    def put(self, batch):
+        self._ids.add(batch.header_signature)
```
If these needs to be kept in sync, atomically, and this and `get` are used across threads, there needs to be a lock around the actions here.  Based on your commit message this doesn't guarantee they stay in sync, from a thread perspective.

aludvik commented:
Fixed a bug I introduced (forgot to reset the pending batches pool in one place) and added a new commit to prevent blocks from being built outside the publisher thread.

peterschwarz commented on validator/sawtooth_validator/journal/publisher.py:
```diff
@@ -427,10 +433,10 @@ def finalize_block(self, identity_signer, pending_batches):
+    example it tracks 'consensus_ready', which denotes entering or exiting of
+    that state."""
+    def __init__(self, consensus_ready=True):
+        self.consensus_ready = True
```
Clearly, the old version was the same, but the parameter is not being used.

aludvik commented:
Completed overnight stability check of 10 node PoET network under 20 TPS workload. Network stayed in consensus and did 9.88 TPS.
